Add 'A pivotal summer' blog post for Trino 483 (#841)
trinodb/trino.io: _posts/2026-07-18-a-pivotal-summer.md

---
layout: post
title: "A pivotal summer"
author: "Martin Traverso"
excerpt_separator: <!--more-->
image: /assets/blog/a-pivotal-summer.png
---

> Last month we called Trino 482 the [summer of grammar](/blog/2026/06/26/summer-of-grammar.html),

Reviewer: Wording is a bit weird .. maybe 

```suggestion
Last month we called Trino 482 the [summer of grammar release](/blog/2026/06/26/summer-of-grammar.html),
```

Also use the jekyll syntax

```
[Announcement blog post]({{ site.baseurl }}{% post_url 2020-04-11-the-definitive-guide %})
```

because it closed so many small gaps in Trino's dialect of SQL at once. That

> post was mostly about breadth: a long list of standard predicates and forms,

Reviewer: Not the post .. the list of features and enhancement was the list .. maybe reword along those lines

each a modest convenience on its own.

> Trino 483 keeps the season going, but the character is different. Instead of a

Reviewer: Instead ... is a really long weird sentence. Break it up.

dozen small gaps, it lands a handful of larger, more powerful additions, the kind
that change the *shape* of the queries you write rather than just tidying them
up. One of them, `PIVOT`, is big enough to carry the release on its own, so

> that's where we'll start and where we'll linger longest. As before, every example

Reviewer: remove the "so that's where we'll start and where we'll linger longest"

> is live: hit **Run** and watch Trino 483 evaluate it for real.

Reviewer: ```suggestion
is live. Hit **Run** and watch Trino 483 evaluate it.
```

<!--more-->

<script src="https://trysql.io/embed.js" async></script>

<style>
try-sql { display: block; margin: 1.5rem auto; }
try-sql iframe { display: block; margin: 0 auto; }
</style>

## `PIVOT`, at last

> The headline feature is the [`PIVOT`](/docs/current/sql/pivot.html) clause. If

Reviewer: add
```suggestion
The headline feature is the [`PIVOT`]({{ site.baseurl }}/docs/current/sql/pivot.html) clause. If
```

> you've ever reached for a pile of `CASE` expressions wrapped in aggregates to

Reviewer: maybe written or needed instead of "reached"

turn a categorical column into a column *per category*, this is for you.

Every example in this section runs against the same little `sales` table: one row
per region, channel, and month, with an amount. To see why `PIVOT` is one of the
longest-standing feature requests, start with what you had to write before: one
`sum(CASE ...)` per month, spelled out by hand, with the column names left as
your own bookkeeping.

> <try-sql version="483" height="320">

Reviewer: Does the version here declare a minimum or an exact value? Will this work in the future?

<pre data-query>SELECT region,
       sum(CASE WHEN month = 1 THEN amount END) AS jan_total,
       sum(CASE WHEN month = 2 THEN amount END) AS feb_total
FROM (VALUES
    ('east', 'web',   1, 100),
    ('east', 'store', 1,  40),
    ('east', 'web',   2, 150),
    ('west', 'web',   1, 200),
    ('west', 'store', 1,  60),
    ('west', 'store', 2,  90)
) AS sales(region, channel, month, amount)
GROUP BY region</pre>
</try-sql>

> `PIVOT` says the same thing declaratively. You give it one or more aggregations,

Reviewer: ```suggestion
`PIVOT` says the same thing declaratively. You provide one or more aggregations,
```

a pivot column whose values become the new columns, and the list of values you
care about. Each additional month is just another entry in the `IN` list rather
than another hand-written `CASE`:

<try-sql version="483" height="340">
<pre data-query>SELECT *
FROM (VALUES
    ('east', 'web',   1, 100),
    ('east', 'store', 1,  40),
    ('east', 'web',   2, 150),
    ('west', 'web',   1, 200),
    ('west', 'store', 1,  60),
    ('west', 'store', 2,  90)
) AS sales(region, channel, month, amount)
PIVOT (
    sum(amount) AS total
    FOR month IN (1 AS jan, 2 AS feb)
    GROUP BY region
)</pre>
</try-sql>

The output has columns `region`, `jan_total`, and `feb_total`. Each pivoted
column name is stitched together from the pivot value's alias and the
aggregation's alias, and the `GROUP BY` names the dimensions that survive as
ordinary columns.

### More than one number per cell

A `PIVOT` can compute several aggregations at once. When it does, each one needs
an alias, so the pivot can build unambiguous column names by pairing every value

> with every aggregation. Here each month gets both a total and an order count:

Reviewer: ```suggestion
with every aggregation. In the following example each month gets both a total amount and an order count:
```

<try-sql version="483" height="360">
<pre data-query>SELECT *
FROM (VALUES
    ('east', 'web',   1, 100),
    ('east', 'store', 1,  40),
    ('east', 'web',   2, 150),
    ('west', 'web',   1, 200),
    ('west', 'store', 1,  60),
    ('west', 'store', 2,  90)
) AS sales(region, channel, month, amount)
PIVOT (
    sum(amount) AS total,
    count(*) AS orders
    FOR month IN (1 AS jan, 2 AS feb)
    GROUP BY region
)</pre>
</try-sql>

The result gains `jan_total`, `jan_orders`, `feb_total`, and `feb_orders`: the

> columns for each value stay grouped together, in the order the aggregations were

Reviewer: ```suggestion
columns for each value stay grouped together, in the order the aggregations are
``a

becomes its own column. This is how you build a cell per (channel, month) pair:

<try-sql version="483" height="360">
<pre data-query>SELECT *
FROM (VALUES
    ('east', 'web',   1, 100),
    ('east', 'store', 1,  40),
    ('east', 'web',   2, 150),
    ('west', 'web',   1, 200),
    ('west', 'store', 1,  60),
    ('west', 'store', 2,  90)
) AS sales(region, channel, month, amount)
PIVOT (
    sum(amount) AS total
    FOR (channel, month) IN (
        ('web', 1) AS web_jan,
        ('web', 2) AS web_feb,
        ('store', 1) AS store_jan
    )
    GROUP BY region
)</pre>
</try-sql>

Notice `west`'s `web_feb_total` comes back `NULL`: no row matches that
combination, so the aggregate sees an empty input. That is the one sharp edge to
remember. The `IN` list is a fixed set of constants you write out, so `PIVOT`
only produces columns for the values you name, and a value that never appears in
the data still produces a column full of empty-input results.

Because the result of a `PIVOT` is itself just a relation, it can be aliased,
nested in a subquery, or fed straight into another `PIVOT`.

## Navigating JSON without the ceremony

> Trino has always had powerful tools for querying JSON: [`json_query`,

Reviewer: Either remove the : after JSON or make it a proper list

> `json_value`, and `json_exists`](/docs/current/functions/json.html). But they

Reviewer: ```suggestion
`json_value`, and `json_exists`](/docs/current/functions/json.html). Unfortunately they
```

or leave it as one sentence with the first part if you use "but"

ask you to write out a full JSON path expression and a `RETURNING` clause for
even the simplest lookup. The SQL specification also defines a more convenient

> [simplified accessor](/docs/current/functions/json.html) syntax, and Trino 483

Reviewer: baseurl

Reviewer: ```suggestion
[simplified accessor](/docs/current/functions/json.html) syntax. Trino 483
```

> now implements it, letting you reach into a `json` value with the same dotted

Reviewer: ```suggestion
now implements it and lets you reach into a `json` value with the same dotted
```

and subscripted notation you'd use on a row or an array.

The receiver has to be a value of the `json` type. Below, a `JSON` literal

> supplies the document. (Watch out for one trap: `CAST('...' AS json)` does *not*

Reviewer: ```suggestion
supplies the document. Watch out for one trap: `CAST('...' AS json)` does *not*
```

I would remove the brackets

`j.Foo` and `j.foo` are different paths, unlike ordinary SQL identifiers. The
item-method names, on the other hand, are regular SQL identifiers and stay
case-insensitive.

There's also a wildcard form. `SELECT j.*` expands to the top-level members of a
JSON object, returned as a single column holding a JSON array:

<try-sql version="483" height="200">
<pre data-query>SELECT j.* AS (members)
FROM (VALUES JSON '{"a": 1, "b": 2, "c": 3}') t(j)</pre>
</try-sql>

## Parsing dates inside JSON paths

Staying with JSON for a moment: SQL/JSON path expressions gained a `datetime()`
method, which parses a textual item into a proper date or time value from inside
the path itself. It takes an optional format template, so you can handle the
non-ISO date strings that turn up in real-world documents without post-processing
in SQL:

<try-sql version="483" height="200">
<pre data-query>SELECT json_value('{"order_date": "17-07-2026"}',
                  'lax $.order_date.datetime("DD-MM-YYYY")'
                  RETURNING date) AS order_date</pre>
</try-sql>

Without a format template, `datetime()` reads the value as the most specific of

> `date`, `time`, or `timestamp` (with or without a time zone) that its shape

Reviewer: ```suggestion
`date`, `time`, or `timestamp` with or without a time zone that its shape
```

naive `start1 <= end2 AND start2 <= end1` doesn't say much about intent. The
standard [`OVERLAPS`](/docs/current/functions/datetime.html) predicate now
expresses it directly. Each operand is a `(start, end)` pair, and the result is
`true` when the two periods share any instant:

<try-sql version="483" height="220">
<pre data-query>SELECT (DATE '2026-06-01', DATE '2026-08-31')
           OVERLAPS (DATE '2026-08-01', DATE '2026-09-30') AS summer_meets_fall,
       (DATE '2026-01-01', DATE '2026-03-31')
           OVERLAPS (DATE '2026-06-01', DATE '2026-08-31') AS q1_meets_summer</pre>
</try-sql>

The second element of a pair can also be an `interval` instead of an end point,
in which case the end is computed as `start + interval`, handy when you know a
period by its length rather than its endpoints. `OVERLAPS` uses half-open
semantics, so two periods that merely touch at a boundary don't count as
overlapping. And since no corner of SQL escapes the question of `null`, a `null`
endpoint isn't an error: it simply leaves that side of the period open-ended.

## Geometry grows a third dimension

The language additions have plenty of company. Trino 483 substantially expands

> the [geospatial](/docs/current/functions/geospatial.html) function library, and

Reviewer: baseurl

</try-sql>

The second is **reprojection**. `ST_Transform` (and `ST_TransformXY`, which
leaves Z untouched) converts a geometry between coordinate reference systems by
EPSG SRID, so you can move between, say, WGS 84 longitude/latitude and Web
Mercator. Paired with the new `ST_GeomFromEWKT`, which parses well-known text
with an `SRID=` prefix, projecting a point takes one expression:

<try-sql version="483" height="220">
<pre data-query>SELECT ST_AsText(
    ST_Transform(
        ST_GeomFromEWKT('SRID=4326;POINT (-71.0882 42.3607)'),
        3857)) AS web_mercator</pre>
</try-sql>

Beyond those, there's a whole batch of new constructors and operations:
`ST_MakeLine`, `ST_Collect`, `ST_Polygonize`, `ST_VoronoiPolygons`,
`ST_MinimumBoundingCircle`, `ST_OrientedEnvelope`, and the
`geometry_collect_agg` aggregate, among others. Here's a line assembled from a
handful of points:

<try-sql version="483" height="220">
<pre data-query>SELECT ST_AsText(
    ST_MakeLine(ARRAY[ST_Point(0, 0), ST_Point(1, 1), ST_Point(2, 0)])) AS line</pre>
</try-sql>

## A small one to finish

Not every addition needs to reshape your queries. There's a new

> [`title_case`](/docs/current/functions/string.html) function that capitalizes

Reviewer: haha .. now we need sentence case and camel case and snake case and kebab case

the first letter of each word and lowercases the rest, the obvious companion to
the `upper` and `lower` you already know:

<try-sql version="483" height="200">
<pre data-query>SELECT title_case('the trino summer of grammar') AS titled</pre>
</try-sql>

## Wrapping up

Where 482 was about filling in the many small corners of standard SQL, 483 is

> about a few genuinely new capabilities: pivoting rows into columns, navigating

Reviewer: Make it an actual list that stands out

```suggestion
about a few genuinely new capabilities: 

* Pivoting rows into columns
* Navigating ...
```

JSON as if it were native, and asking a straightforward question about
overlapping time. Different flavor, same goal: a dialect of SQL that lets you
say what you mean with less ceremony.

And that's just the language side of the ledger. Trino 483 also ships `char`
column filter fixes, a redesigned Web UI now serving as the default, and a long
list of connector improvements across Delta Lake, Iceberg, Hive, and more.
Thirsty for the rest? The [Trino 483 release

> notes](/docs/current/release/release-483.html) have the full accounting.

Reviewer: baseurl
